eclass/dlang-compilers-r1.eclass

```diff
@@ -0,0 +1,291 @@
+# Copyright 2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+# @ECLASS: dlang-compilers-r1.eclass
+# @MAINTAINER:
+# Andrei Horodniceanu <a.horodniceanu@proton.me>
+# @AUTHOR:
+# Andrei Horodniceanu <a.horodniceanu@proton.me>
+# Based on dlang-compilers.eclass by Marco Leise <marco.leise@gmx.de>.
+# @BUGREPORTS:
+# Please report bugs via https://github.com/gentoo/dlang/issues
+# @VCSURL: https://github.com/gentoo/dlang
+# @BLURB: Support data for dlang-utils.eclass
+# @DESCRIPTION:
+# Contains the available D compiler versions with their stable
+# architectures and the language version they support.
+
+if [[ ${_DLANG_COMPILERS_R1_ECLASS} != "recur -_+^+_- spank" ]] ; then
+_DLANG_COMPILERS_R1_ECLASS="recur -_+^+_- spank"
+
+# @ECLASS_VARIABLE: _DLANG_DMD_FRONTENDS
+# @INTERNAL
+# @DESCRIPTION:
+# A list of dmd implementations with their dlang frontend version (which
+# happens to coincide with the implementation version) and the
+# architectures they support as a list of keywords. Only stable and
+# unstable keywords will appear.
+#
+# The elements are ordered, the higher implementation version has the
+# higher index.
+#
+# Example value for an element:
+# @CODE
+# "2.102 2.102 ~amd64 x86"
+# @CODE
+# Where the first 2.102 represents the implementation version,
+# the second one represents the D language version the implementation
+# supports and, lastly, there are two keywords.
+readonly _DLANG_DMD_FRONTENDS=(
+	"2.101 2.101 ~amd64 ~x86"
+	"2.102 2.102 ~amd64 ~x86"
+	"2.103 2.103 ~amd64 ~x86"
+	"2.104 2.104 ~amd64 ~x86"
+	"2.105 2.105 ~amd64 ~x86"
+	"2.106 2.106 ~amd64 ~x86"
+	"2.107 2.107 ~amd64 ~x86"
+)
+
+# @ECLASS_VARIABLE: _DLANG_GDC_FRONTENDS
+# @INTERNAL
+# @DESCRIPTION:
+# A list of gdc implementations with their dlang frontend version and
+# the architectures they support as a list of keywords. Only stable and
+# unstable keywords will appear.
+#
+# The elements are ordered, the higher implementation version has the
+# higher index.
+#
+# Example value for an element:
+# @CODE
+# "13 2.103 amd64 ~arm64 x86"
+# @CODE
+# Where 13 represents the implementation version, 2.103 represents the D
+# language version the implementation supports and, lastly, there are
+# three keywords.
+readonly _DLANG_GDC_FRONTENDS=(
```

Comment:
KEYWORDS in an eclass is a code smell and makes doing some things hard later on. See https://bugs.gentoo.org/342185.

Reply:
I think you've misunderstood what I refer to as *keyword* in the context of this (and the other 2) arrays. I call it keyword since it looks like one but it server a purpose similar to `profiles/`. If you think as the implementation use flags as being `use.mask`ed on all profiles, then, a keyword being present in that list means that the USE flag is unmasked for the corresponding architecture. If the keyword is unstable it would mean that it is *kinda* `use.stable.mask`ed. The usage of these values becomes problematic in the functions below, that one I agree with.

Reply:
But the array is talking about `amd64` and such which is why it looks suspicious. Encoding anything about this and acting on it is suspicious.

Reply:
Maybe I'm misreading your comment but your main concern is still the acting on these values part, not the way that these values are represented in the array, right?

Reply:
I think it implies an abstraction issue if we're doing anything storing or acting on those values. But let me read/digest/consider https://github.com/gentoo/dlang/pull/130#discussion_r1532088163 and answer over there.